caluma/caluma_form/structure.py

```diff
@@ -0,0 +1,212 @@
+"""Hierarchical representation of a document / form."""
+import weakref
+from functools import singledispatch
+from typing import Optional
+
+from .models import Question
+
+
+def object_local_memoise(meth):
+    def new_meth(self, *args, **kwargs):
+        if not hasattr(self, "_memoise"):
+            self._memoise = {}
+
+        key = str([args, kwargs, meth])
+        if key in self._memoise:
+            return self._memoise[key]
+        ret = meth(self, *args, **kwargs)
+        self._memoise[key] = ret
+        return ret
+
+    return new_meth
```

Comment:
I always prefer longer variable names if the abbreviation is not completely obvious. This is an extreme example where the abbreviation even has a (very different) meaning :wink:

Reply:
Also, doesn't this break the method depends on `self`?

Reply:
Fixed the naming. And no, it is explicitly meant to be used on methods (vs functions), so it's fine to depend on `self`

Reply:
I meant more if that state of the object changes, the memoisation can return incorrect results because the cache key doesn't depend on the state of the object.

Reply:
Ah. In that case you'd be right, but in this situation, the objects will be virtually immutable, so it should be fine

Reply:
:+1:

Reply:
![Walter-Whites-Smooth-Wink-On-Breaking-Bad](https://user-images.githubusercontent.com/6150577/73455037-98ab1080-436f-11ea-945e-eb49c9bc96ae.gif)